### PR TITLE
Visualize actual domain boundary of parallel meshes

### PR DESCRIPTION
**Pull request by HennesHajduk**
Here's my suggested fix regarding

https://github.com/GLVis/glvis/issues/223#issue-1174689847

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2961](https://github.com/mfem/mfem/pull/2961) | @HennesHajduk | @tzanio | @jandrej + @tzanio | 4/11/22 | 4/13/22 | 4/14/22 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Review by tzanio — APPROVED** [review]

**Review by jandrej — APPROVED** [review]

**Comment by tzanio** [comment]
Unfortunately this doesn't quite work because clang complains that `Print` is a virtual method and we are adding the second  (optional) parameter only in the derived class :(

**Comment by jandrej** [comment]
Um, what about a
```
using Mesh::Print;
```
in `ParMesh`?

**Comment by tzanio** [comment]
> https://godbolt.org/z/zKqq8E9xr

This doesn't seem to work correctly with the virtual function.

```c++
    {
        Mesh *m;
        m = new Mesh; 
        m->Print(std::cout);
        m = new ParMesh; 
        m->Print(std::cout);
    }
```

**Comment by tzanio** [comment]
I see 3 options to resolve this

1️⃣  We add a fake optional `bool` parameter to `Mesh::Print` so the two methods match.

2️⃣  We rename `ParMesh::Print` to something else e.g. `ParMesh::Printer`, `ParMesh::PrintVis` or `ParMesh::PrintWithShared` and set `ParMesh::Print` to have the old signature and just call `Printer(out, true)`

3️⃣  We add `print_shared` as a data member of `ParMesh` and add a method, `ParMesh::SetPrintShared`, to set it.

Which option do you prefer? Any other suggestions?

**Comment by HennesHajduk** [comment]
> three We add `print_shared` as a data member of `ParMesh` and add a method, `ParMesh::SetPrintShared`, to set it.

Seems like the best way to deal with it IMO.

**Comment by tzanio** [comment]
@HennesHajduk and @jandrej -- how does https://github.com/mfem/mfem/pull/2961/commits/a59deb867527de4236aed480733b2578b629bcc8 look?

File: mesh/pmesh.hpp [hunk 1]
```diff
@@ -84,6 +84,10 @@ class ParMesh : public Mesh
    mutable long glob_elem_offset, glob_offset_sequence;
    void ComputeGlobalElementOffset() const;
 
+   // Should Print() add the parallel interface as boundary (typically used for
+   // visualization purposes)?
```
Comment (jandrej) [inline mesh/pmesh.hpp:88]: Never write documentation with a question.

```// Enable Print() to add the parallel interface as boundary (typically used for visualization purposes)```

File: mesh/pmesh.cpp [hunk 2]
```diff
@@ -1126,6 +1133,7 @@ void ParMesh::LoadSharedEntities(istream &input)
 ParMesh::ParMesh(ParMesh *orig_mesh, int ref_factor, int ref_type)
 {
    MakeRefined_(*orig_mesh, ref_factor, ref_type);
+   print_shared = true;
```
Comment (jandrej) [inline mesh/pmesh.cpp:1136]: Use default initializer in the header file for default initialization.

File: mesh/pmesh.cpp [hunk 3]
```diff
@@ -936,6 +941,8 @@ ParMesh::ParMesh(MPI_Comm comm, istream &input, bool refine)
    const int gen_edges = 1;
 
    Load(input, gen_edges, refine, true);
+
+   print_shared = true;
```
Comment (jandrej) [inline mesh/pmesh.cpp:945]: Use default initializer in the header file for default initialization.

File: mesh/pmesh.cpp [hunk 4]
```diff
@@ -863,6 +867,7 @@ ParMesh::ParMesh(const ParNCMesh &pncmesh)
    Mesh::InitFromNCMesh(pncmesh);
    ReduceMeshGen();
    have_face_nbr_data = false;
+   print_shared = true;
```
Comment (jandrej) [inline mesh/pmesh.cpp:870]: Use default initializer in the header file for default initialization.

File: mesh/pmesh.cpp [hunk 5]
```diff
@@ -296,6 +298,8 @@ ParMesh::ParMesh(MPI_Comm comm, Mesh &mesh, int *partitioning_,
    }
 
    have_face_nbr_data = false;
+
+   print_shared = true;
```
Comment (jandrej) [inline mesh/pmesh.cpp:302]: Use default initializer in the header file for default initialization.

File: mesh/pmesh.cpp [hunk 6]
```diff
@@ -91,6 +91,8 @@ ParMesh::ParMesh(const ParMesh &pmesh, bool copy_nodes)
       *Nodes = *pmesh.Nodes;
       own_nodes = 1;
    }
+
+   print_shared = true;
```
Comment (jandrej) [inline mesh/pmesh.cpp:95]: Use default initializer in the header file for default initialization.

Comment (tzanio) [reply mesh/pmesh.cpp:95]: Good point, can you make this change?

There are probably other places in MFEM that can be simplified with default member initialization.

More generally, we should go through the code and take advantage of C++11 features like this where it makes sense.

File: mesh/pmesh.hpp (hunk 1 above)
Comment (tzanio) [reply mesh/pmesh.hpp:88]: I am not sure I agree that this is more clear, but I am OK with changing it

**Comment by HennesHajduk** [comment]
LGTM

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**